### ISSUE TITLE
A dropped NWP grid point silently lands as a physically-plausible zero, not null

### ISSUE THREAD

**Issue by JackKelly** [openclimatefix/nged-substation-forecast#477]
Surfaced during review of PR #468 (#161), and deliberately left out of that PR's scope — but it's
a silent data-correctness bug, not a documentation gap, so it needs its own fix rather than staying
only-documented.

This issue is for a **quick, temporary fix only**, worth having in place ASAP because of how
quietly this can corrupt ML training data. The permanent, "proper" fix is tracked separately as a
child of #323 (referenced there, and it references back here).

## The bug

`_process_chunk_for_1_lead_time_and_1_ens_member` in
`packages/dynamical_data/src/dynamical_data/ecmwf_ens/convert_to_polars.py` left-joins the H3 grid
weights onto the raw NWP grid, then aggregates each H3 cell as a proportion-weighted `.sum()`:

```python
joined = h3_grid.join(
    nwp_df, left_on=["nwp_lon", "nwp_lat"], right_on=["longitude", "latitude"], how="left"
)
...
joined.with_columns(pl.col(numeric_vars) * pl.col("proportion"))
    .group_by("h3_index")
    .agg(pl.col(numeric_vars).sum(), ...)
```

If **every** raw grid point that contributes to one H3 cell is absent from the source dataset (a
join miss — e.g. a lat/lon `h3_grid_weights` expects isn't present in that day's `ds`), every value
going into that cell's group is `null`. Polars sums an all-null `Float32` group to `0.0`, not to
`null` — confirmed empirically:

```python
>>> pl.DataFrame({"g": [1, 1], "v": [None, None]}, schema={"g": pl.Int64, "v": pl.Float32}) \
...     .group_by("g").agg(pl.col("v").sum(), n=pl.col("v").count())
shape: (1, 3)
┌─────┬─────┬─────┐
│ g   ┆ v   ┆ n   │
│ i64 ┆ f32 ┆ u32 │
╞═════╪═════╪═════╡
│ 1   ┆ 0.0 ┆ 0   │
└─────┴─────┴─────┘
```

The existing `.fill_nan(None)` call after aggregation does nothing here — `0.0` is not `NaN`, it's
a plausible-looking real value. So a dropped grid point lands as **0 °C / 0 Pa / 0 m s⁻¹**, inside
every physical bound `Nwp` declares, and is invisible to all three of: `Nwp.validate`, the existing
`nwp_has_no_unexpected_nulls` check, and the new `nwp_run_is_complete` check from #161/PR #468. It
would only ever surface as an unexplained outlier in downstream model behaviour — exactly the
"invalidates the training data without anyone noticing" failure mode.

I have **not** confirmed this has actually fired against real production data yet — only that the
code path is reachable and reproduces the bad value on an injected join-miss. Worth checking
whether any real historical NWP partition has an anomalous exact-zero cell before assuming it's
purely theoretical.

## Why a naive "assert the cell count matches" fix does not work

It's tempting to reach for the same fix as the completeness check — assert the output has exactly
the expected number of H3 cells — but that's a no-op here. The join is **from** `h3_grid` (the left
side), so every `h3_index` in `h3_grid_weights` survives the join and lands in exactly one
`group_by` group regardless of whether any of its raw grid points matched. The row/cell count is
already structurally invariant under this bug — that's precisely why
`docs/architecture/ecmwf-ens-known-issues.md` notes the completeness check's cell-count and
row-count checks "cannot fire through today's converter." The corruption is a **wrong value inside
an existing row**, not a missing row, so nothing that counts rows will ever catch it.

## The quick, temporary fix

**Suggested fix:** in the `group_by(...).agg(...)` step, also aggregate a per-variable
contributor count (`pl.col(v).count()`, which Polars already excludes nulls from) and null out any
`numeric_vars` cell whose count is `0`, instead of leaving Polars' default `0.0` sum:

```python
.agg(
    pl.col(numeric_vars).sum(),
    pl.col(numeric_vars).count().name.suffix("_n_contrib"),
    ...
)
.with_columns(
    pl.when(pl.col(f"{v}_n_contrib") == 0).then(None).otherwise(pl.col(v)).alias(v)
    for v in numeric_vars
)
```

This needs no new detection code: once the value is genuinely `null`, the **existing**
`Nwp.validate` checks do the rest — non-nullable instantaneous variables get rejected by base
Patito validation, and a whole-slice null in a de-accumulated variable beyond lead-0 gets rejected
by `_check_no_whole_null_deaccumulated_slices`. The whole run fails to ingest rather than landing
silently corrupted, which is the deliberately blunt, deliberately temporary trade-off: it converts
"invisible corruption" into "a loud missed run," at the cost of failing the entire partition over
what might be a single dropped grid point. That's a known, deliberate, temporary departure from
"never raise on absent input" ([Inherent Stability](https://openclimatefix.github.io/nged-substation-forecast/design-philosophy/inherent-stability/)) —
accepted here because silent corruption is worse than a loud missed run, until the permanent fix
lands.

**Note this fix only catches total dropout** — an H3 cell whose *every* contributing grid point is
missing. A cell fed by multiple grid points at different weights (see
`test_convert_proportion_weighted_aggregation` in
`packages/dynamical_data/tests/test_convert_to_polars.py`) where only *some* points are missing is
not touched by this fix at all: the weighted sum silently omits the missing points without
renormalizing, so it stays quietly biased low. That softer version of the same bug is out of scope
here — see the permanent-fix issue.

This function carries an existing code comment flagging it as "Convention-sensitive to real ECMWF
ENS data," with an instruction to manually run `uv run pytest --run-network -m network` after any
change — that applies to this fix too, not just the permanent one.

**Comment by JackKelly** [comment]
Checked the open question in the issue body — *"worth checking whether any real historical NWP
partition has an anomalous exact-zero cell before assuming it's purely theoretical"*.

**It has never fired on real data.** Checked via the Delta log's per-file Parquet column statistics
(metadata-only, exact, no data scan), across **all 1,684 files** in the NWP table — i.e. the whole
table, all 842 `init_time` partitions from 2024-04-01 to 2026-07-21:

| Column | Global min | Files with `min == 0.0` | Files with null stats |
|---|---:|---:|---:|
| `pressure_surface` | 85,408 Pa | 0 | 0 |
| `pressure_reduced_to_mean_sea_level` | 92,464 Pa | 0 | 0 |
| `geopotential_height_500hpa` | 4,753 m | 0 | 0 |

These three are the right detectors because an exact `0.0` is *physically impossible* for them,
unlike `temperature_2m` where 0 °C is an ordinary value. The all-null-group-sums-to-zero failure in
this issue zeroes **every** numeric variable for the affected cell simultaneously, so any occurrence
would show up as a zero in all three.

Coverage is complete rather than sampled: the stats are non-null for every one of the 1,684 files,
so there is no file where the check silently had nothing to look at.

**Consequences for the fix:**

- No existing training data is corrupted, so this needs no data repair and **no model retrain** —
  it is purely a forward-looking guard. (Relevant to #484, which is closed on that basis.)
- The bug is still worth fixing exactly as scoped. The code path is reachable and reproduces the
  bad value on an injected join-miss; the reason it hasn't bitten is that no join miss has occurred
  yet, not that one can't.

**Comment by JackKelly** [comment]
*Written by Claude Code, at Jack's direction, from the measurement work in
[PR #516](https://github.com/openclimatefix/nged-substation-forecast/pull/516).*

This issue is fixed by PR #516 and will close when that merges. One measurement from it is worth
recording here, because it corrects the mental model in this issue's body.

## `main`'s null cells were mostly self-inflicted

This issue describes the all-null group summing to `0.0`. True, but it is only reachable by a
*join miss*. The commoner case upstream — a **NaN** grid point — behaves completely differently on
`main`, and much more destructively:

- NaN **propagates** through `sum`, so a single corrupt grid point turned its *entire* cell NaN,
  which `main`'s trailing `fill_nan(None)` then turned into a null cell.
- A grid point feeds **4.92** H3 cells on average, so one bad pixel killed several cells.

Measured on 2025-06-04 00Z, the worst run in the archive: null cells came out at **2.88× the number
of null grid points**, and 0.014% corrupt `precipitation_surface` grid points produced **4,394**
null cells. Renormalising leaves **339**, none of them newly null — so ~92% of that run's null
cells were amplification, not lost weather.

Archive-wide the whole phenomenon is small: across 862 runs and 6.24 billion rows, only **12 runs**
carry any de-accumulated null beyond the lead-0 floor at all, totalling **6,550 cells**.

## Why that matters for how this issue is remembered

The headline here is a *silent* zero, and that framing drove the original `priority-high`. Two
things narrow it, both measured on `main`:

1. A join miss also nulls `categorical_precipitation_type_surface`, which
   `Nwp._check_variables_that_were_introduced_after_start_of_dataset` rejects for
   `init_time > 2024-11-12`. So the silent zero only ever bit pre-2024-11-13 backfills; a modern run
   failed loudly, just with the wrong variable named. `priority-high` was removed on that basis.
2. The much likelier NaN path never produced a zero at all — it produced an over-broad null.

So the real value delivered here is **accuracy and diagnosis**, not the prevention of silent
corruption. The fix is still worth having: the estimator was biased low whenever any point was
missing, which is the [#496](https://github.com/openclimatefix/nged-substation-forecast/issues/496)
half, and a genuinely uncovered cell is now null instead of a plausible zero.
